pkg/destroy/alibabacloud/alibabacloud.go

```diff
@@ -1248,6 +1248,7 @@ func (o *ClusterUninstaller) deleteDNSRecords() (err error) {
 
 	// Get the base domain from the cluster domain. the format of cluster domain is '<cluster name>.<base domain>'.
 	baseDomain := strings.Join(strings.Split(o.ClusterDomain, ".")[1:], ".")
+	rr := fmt.Sprintf("api.%s", strings.Split(o.ClusterDomain, ".")[0])
```

Comment:
You could consider using a structured logger here. See the example on the destroy flow for Route53 in AWS:
https://github.com/openshift/installer/blob/06e3fe64f3616f6d6baff95dde5cd3389cf785aa/pkg/destroy/aws/aws.go#L1824-L1832

The `logger` for this function is being used in every function that will call, tracking the id of hosted zone.

Reply:
``` go
rr := fmt.Sprintf("api.%s", strings.Split(o.ClusterDomain, ".")[0])
```
@kwoodson @staebler I am also not sure if we should delete all RRs with a suffix of the `o.ClusterDomain`, rather than only the `api.*`, I guess we will have a leak of RR `*.apps.(o.ClusterDomain)` using this approach. wdyt?

Reply:
@bd233 I tested this code and the code from the #5411.

The domain I was using was `*.apps.test` (created by the ingress controller) as well as `api.test` (created by the installer). I created another domain to verify that it was not removed (`*.apps.test1` and `api.test1`). When calling the `destroy` cluster I verified that `api.test` was removed. This is an improvement but I would also like to see the record from the ingress removed `*.apps.test`. We do not want to leak these resource records.

I *think* we will want to remove all resource records with the `o.ClusterDomain`. That should be relatively safe since the cluster domain should be unique.

Reply:
Not exactly. We cannot guarantee that the cluster domain is unique. What we do for AWS, which I recommend following here as well, is that we only delete the public DNS record if there is also a private DNS record for the same domain.
See https://github.com/openshift/installer/blob/d3736c2e9e10b0b8e5a0ecbf666ff2f739a02683/pkg/destroy/aws/aws.go#L1832-L1865

Reply:
Sorry, I don't understand why the cluster domain is not unique. Can you provide more details. Can users create two clusters with the same cluster name and base domain (I think this is not feasible)? Or consider that the user may manually add the records containing the cluster domain？

Reply:
The installer does not know about any other clusters that exist. So the installer cannot directly verify that someone has not attempted to install a cluster with the exact same cluster domain. Moreover, the destroyer must be resilient to any changes that have been made in the cloud account in between when the installation attempt occurred and when the destroyer is called. The destroyer should always err on the side of caution because the worst thing that it can do is to delete a resource that did not actually belong to the cluster.

Let's look at an example. The user attempts to create a cluster with the cluster domain test.example.com. The installation failed at some point before creating the public DNS records. The next day, the user fixes whatever was causing the installation to fail the first time and tries another installation with the same cluster domain. This installation was successful. The next day, the user realizes that there are some leftover resources from the first installation attempt because the user forgot to destroy the cluster. The user runs the destroyer on the partial cluster from the first attempt. The destroyer sees that there is a *.apps.test.example.com public DNS record, so the destroyer deletes it. Now the user cannot access the apps on their working cluster because the public DNS record has been deleted.